### ISSUE TITLE
prev() issue when cron expression contains 59 seconds

### ISSUE THREAD

**Issue by dianalaza** [harrisiirak/cron-parser#178]
Invoking prev() on a cron expression that have set seconds = 59 does not work ok. At every other invocation, only milliseconds are changed. Tested on 2.13.0
next() is working fine.

`const cronParser = require("cron-parser");

 let cron = cronParser.parseExpression("59 59 23 * * *");
 let i=0;
 while(i++<100){
        Logger.debug(`prev()= ${cron.prev().toISOString()}`);
 }
[Debug] prev()= 2020-01-08T21:59:59.999Z 
[Debug] prev()= 2020-01-08T21:59:59.000Z 
[Debug] prev()= 2020-01-07T21:59:59.999Z
[Debug] prev()= 2020-01-07T21:59:59.000Z
[Debug] prev()= 2020-01-06T21:59:59.999Z
[Debug] prev()= 2020-01-06T21:59:59.000Z`

**Comment by harrisiirak** [comment]
@dianalaza thanks for reporting this! Definitely a bug. I'll look into this as soon as possible.

**Comment by bazineta** [comment]
I don't have a solution yet, but this seems to have been introduced by the fix to #98. That is, reverting the changes for that issue causes a test case for this issue to pass. Breaks the test case for that issue, as expected. It's seemingly got to be something right around here, though:
https://github.com/harrisiirak/cron-parser/blob/f36281a84612b84395b1b275c7e74a2e5e9b83f1/lib/expression.js#L628

**Comment by bazineta** [comment]
I feel as if I'm getting closer to understanding this, but suspect that @santigimeno might find the problem to be more obvious than me.

**Comment by bazineta** [comment]
Well, at least I understand it now.

At around line 628, we're in the handling for having been through a loop iteration and finding that nothing changed. Something has to change or we'll be there forever, and prior to #98, the code moved a second in the appropriate direction.

The fix for #98 avoids subtracting by a complete second when going in the reverse direction because that would be too far a jump for the #98 test case. That test case is a seconds step of */5, so it's due every 5 seconds, i.e., the seconds field will contain `[0, 5, 15, 20, 25, 30, 35, 40, 45, 50, 55]`. The `currentDate` in the #98 test case is set to 25.002 seconds, and we're moving in the reverse direction, so the next stop is 25; if it were to subtract a full second, it'd miss that stop and instead return 20.

The #98 fix therefore just zeroes the milliseconds when going backward, instead of subtracting a second. Unfortunately, that means the next pass through the loop matches on seconds, so you get the behavior described here; #98 passes, but in this case, it returns the same stop twice.

**Comment by bazineta** [comment]
I think this might actually be due to the way subtraction is handled in `date.js`. In month, day, hour and minute subtraction, the milliseconds are going to end up being 999. In second subtraction, they're going to be 0.

I think it follows that if they're always set to 0, i.e., append a `.startOf('second');` to the end of the month, day, hour and minute subtraction functions, this might just work as expected. The test cases do all pass, without any changes to `expression.js`, including a new test case for this issue and the #98 test case.

**Comment by bazineta** [comment]
Test case:
```
test('it will work with #178 issue case', function(t) {
  var interval = CronExpression.parse('59 59 23 * * *',{
    currentDate : new Date('2012-12-26 14:38:53')
  });

  [25, 24, 23, 22].forEach(function(date) {
    var prev = interval.prev();
    t.equal(prev.getFullYear(), 2012);
    t.equal(prev.getMonth(), 11);
    t.equal(prev.getDate(), date);
    t.equal(prev.getHours(), 23);
    t.equal(prev.getMinutes(), 59);
    t.equal(prev.getSeconds(), 59);
  });

  t.end();
});
```

Example change to date.js subtraction functions:
```
CronDate.prototype.subtractMonth = function() {
  this._date = this._date.minus({ months: 1 }).endOf('month');
};
```
becomes:
```
CronDate.prototype.subtractMonth = function() {
  this._date = this._date.minus({ months: 1 }).endOf('month').startOf('second');
};
```
Make the same change to the day, hour, and minute subtraction functions. The test case should fail before the change; it should pass after the change, as should all other test cases.

**Comment by harrisiirak** [comment]
@bazineta thanks for debugging this issue, the conclusion and insights are very useful here. The fix you suggested makes sense to me, therefore I created a PR with this fix (#225). Thanks again!
